File: 8_caesar_cipher/caesar_cipher.py
import string
from functools import reduce


def caesar(plain, shift):
    def substitute(acc, letter):
        if not letter.isalpha():
            return acc + letter

        idx = alphabet.find(letter)
        letters = alphabet[:26] if idx < 26 else alphabet[26:]
        return acc + letters[(idx + shift) % 26]

    alphabet = string.ascii_letters

    result = reduce(
        substitute,
        plain,
        ''
    )

    return result


def main():
    assert caesar('This is a caesar cipher example', shift=0) == 'This is a caesar cipher example'
    assert caesar('A not so long string', shift=2) == 'C pqv uq nqpi uvtkpi'
    assert caesar('Negative shift', shift=-1) == 'Mdfzshud rghes'


if __name__ == '__main__':
    main()

File: 8_caesar_cipher/caesar_cipher.rb
class AssertionError < RuntimeError; end
def assert &block; raise AssertionError unless yield; end

def caesar(plain, shift:)
  alphabets = ['A'..'Z', 'a'..'z'].map do |alphabet|
    Hash[alphabet.map.with_index.to_a]
  end

  plain.chars.map do |char|
    alphabet = alphabets.detect { |alphabet| alphabet.include? char }
    next char unless alphabet
    char_index = alphabet[char]
    new_char_index = (char_index + shift) % alphabet.size
    alphabet.keys[new_char_index]
  end.join
end

assert { caesar('This is a caesar cipher example', shift: 0) == 'This is a caesar cipher example' }
assert { caesar('A not so long string', shift: 2) == 'C pqv uq nqpi uvtkpi' }
assert { caesar('Negative shift', shift: -1) == 'Mdfzshud rghes' }
